narwhal/primary/src/state_handler.rs

```diff
@@ -1,35 +1,185 @@
 // Copyright (c) 2021, Facebook, Inc. and its affiliates
 // Copyright (c) Mysten Labs, Inc.
 // SPDX-License-Identifier: Apache-2.0
-use config::AuthorityIdentifier;
+use config::{AuthorityIdentifier, Committee};
+use crypto::RandomnessPrivateKey;
+use fastcrypto::groups;
+use fastcrypto_tbls::{dkg, nodes};
 use mysten_metrics::metered_channel::{Receiver, Sender};
 use mysten_metrics::spawn_logged_monitored_task;
+use sui_protocol_config::ProtocolConfig;
+use sui_types::committee::VALIDITY_THRESHOLD;
 use tap::TapFallible;
 use tokio::task::JoinHandle;
 use tracing::{debug, error, info, warn};
-use types::{Certificate, CertificateAPI, ConditionalBroadcastReceiver, HeaderAPI, Round};
+use types::{
+    Certificate, CertificateAPI, ConditionalBroadcastReceiver, HeaderAPI, Round, SystemMessage,
+};
 
-/// Receives the highest round reached by consensus and update it for all tasks.
+type PkG = groups::bls12381::G2Element;
+type EncG = groups::bls12381::G2Element;
+
+/// Updates Narwhal system state based on certificates received from consensus.
 pub struct StateHandler {
-    /// The id of this authority.
     authority_id: AuthorityIdentifier,
+
     /// Receives the ordered certificates from consensus.
     rx_committed_certificates: Receiver<(Round, Vec<Certificate>)>,
     /// Channel to signal committee changes.
     rx_shutdown: ConditionalBroadcastReceiver,
     /// A channel to update the committed rounds
     tx_committed_own_headers: Option<Sender<(Round, Vec<Round>)>>,
+    /// A channel to send system messages to the proposer.
+    tx_system_messages: Sender<SystemMessage>,
+
+    /// If set, generates Narwhal system messages for random beacn
+    /// DKG and randomness generation.
+    randomness_state: Option<RandomnessState>,
 
     network: anemo::Network,
 }
 
+// Internal state for randomness DKG and generation.
+struct RandomnessState {
```

Comment:
I would suggest to add either here or somewhere else some comments on the high level protocol for generating randomness as it involves multiple steps. I think that would help the future reader to better understand the intention.